### PR TITLE
Add ledger_agent, trezor_agent and keepkey_agent packages

### PR DESCRIPTION
**Pull request by hkjn**
###### Motivation for this change

Add new `ledger_agent`, `trezor_agent` and `keepkey_agent` packages, providing the binaries:

* `ledger-agent`
* `ledger-gpg`
* `trezor-agent`
* `trezor-gpg`
* `keepkey-agent`

Also add hkjn as initial maintainer for `pythonPackages.trezor_agent`.

_Edit:_ Updated description after changes from review feedback.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @mmahut @np

### CODE REVIEW

**Comment by mmahut** [comment]
This is the same source code as ```pythonPackages.trezor_agent``` do you mind integrating it from there and while at it, make the trezor-agent as the same way?

**Comment by mmahut** [comment]
And probably ```keepkey``` while we are at it, for consistency.

**Comment by hkjn** [comment]
> This is the same source code as `pythonPackages.trezor_agent` do you mind integrating it from there and while at it, make the trezor-agent as the same way?

Thanks for the feedback and pointer to the existing [pythonPackages.trezor_agent](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/trezor_agent/default.nix), @mmahut! As you say, that package uses the same https://github.com/romanz/trezor-agent source code as is used in this PR in the `ledger_agent`.

I'll rephrase your suggestion to make sure I understood; is the aim to have a `pythonPackages.trezor_agent` define a single python package, that can be referenced in three top-level packages as follows?

- `pkgs/applications/misc/ledgeragent`
- `pkgs/applications/misc/trezoragent`
- `pkgs/applications/misc/keepkeyagent`

I don't have Trezor or Keepkey devices accessible right now, so I can't fully test those changes, but if that's what you were suggesting I can make that change and we can hold off on merging until we can test the change fully.

Also, do you (and/or @np?) want to be added as a maintainer to these top-level packages? (I'm fine not being a maintainer too btw, depending on your preference, just put myself there in this PR to get things started.)

**Comment by mmahut** [comment]
> I'll rephrase your suggestion to make sure I understood; is the aim to have a `pythonPackages.trezor_agent` define a single python package, that can be referenced in three top-level packages as follows?

That is correct.

> I don't have Trezor or Keepkey devices accessible right now, so I can't fully test those changes, but if that's what you were suggesting I can make that change and we can hold off on merging until we can test the change fully.

I think I can help with testing.

> Also, do you (and/or @np?) want to be added as a maintainer to these top-level packages? (I'm fine not being a maintainer too btw, depending on your preference, just put myself there in this PR to get things started.)

Sure, yes, please. More maintainers the better :)

File: pkgs/top-level/all-packages.nix
Comment (worldofpeace) [inline pkgs/top-level/all-packages.nix]:
```suggestion
   ledgeragent = with python3Packages; toPythonApplication ledgeragent;
```

This preferred over adding extra files in the nixpkgs tree.
See other examples in `all-packages.nix`

Comment (hkjn) [reply pkgs/top-level/all-packages.nix]: Thanks, good call. Done.

**Comment by hkjn** [comment]
Thanks for the feedback @mmahut @worldofpeace, please take another look.

> > I'll rephrase your suggestion to make sure I understood; is the aim to have a `pythonPackages.trezor_agent` define a single python package, that can be referenced in three top-level packages as follows?
> 
> That is correct.
> 
I couldn't figure out a way to achieve what you suggested here..

The issue is that the `buildPythonPackage` definition uses `fetchPyPi` to pull down the specific library, one of the ones below depending on which package we're building:

* https://pypi.org/project/trezor-agent/: 0.10.0
* https://pypi.org/project/ledger_agent/: 0.9.0
* https://pypi.org/project/keepkey_agent/: 0.9.0

Because the top-level packages now are just using with `python3Packages; toPythonApplication foo_agent;` (thanks for the suggestion @worldofpeace), I don't see how to avoid having three python packages for the three pip projects we want to pull down.

As an aside, the listed versions above are the latest that have been published, but at https://github.com/romanz/trezor-agent the latest published version is 0.13.1.

cc @romanz, would it be possible to publish the latest release to pypi.org as well?

> 
> > Also, do you (and/or @np?) want to be added as a maintainer to these top-level packages? (I'm fine not being a maintainer too btw, depending on your preference, just put myself there in this PR to get things started.)
> 
> Sure, yes, please. More maintainers the better :)

Cool, I've added the two of you and myself (`hkjn mmahut np`) to all three packages:

* `trezor_agent`
* `ledger_agent`
* `keepkey_agent`

**Comment by romanz** [comment]
> cc @romanz, would it be possible to publish the latest release to pypi.org as well?

The latest versions of specific agents' packages should be on PyPI, albeit under different versions than the `libagent` package (which contains the common library code).
For example, https://pypi.org/project/trezor-agent/ has 0.10.0 - which is the latest TREZOR agent release (see [agents/trezor/setup.py](https://github.com/romanz/trezor-agent/blob/513b1259c4d7aca5f88cd958edc11828d0712f1b/agents/trezor/setup.py#L6)).

Please let me know if I can fix the versioning to be more consistent and easier to follow :)

**Comment by hkjn** [comment]
> > cc @romanz, would it be possible to publish the latest release to pypi.org as well?
> 
> The latest versions of specific agents' packages should be on PyPI, albeit under different versions than the `libagent` package (which contains the common library code).
> For example, https://pypi.org/project/trezor-agent/ has 0.10.0 - which is the latest TREZOR agent release (see [agents/trezor/setup.py](https://github.com/romanz/trezor-agent/blob/513b1259c4d7aca5f88cd958edc11828d0712f1b/agents/trezor/setup.py#L6)).
> 
> Please let me know if I can fix the versioning to be more consistent and easier to follow :)

Ah, that versioning scheme makes sense, given that all four releaseable units (three agents + `libagent`) are in the same repo.

I guess my confusion came from the repo being named `trezor-agent`, but containing the code for all four releases. Perhaps we can add a note to either the Github releases for the repo, or the `README.md`? (I can send a PR, let's continue there.)

**Comment by mmahut** [comment]
This looks good to me, @worldofpeace might confirm is there is a better way to do this.

@1000101 might help with ledger/keepkey testing.

**Review by jonringer — COMMENTED** [review]
otherwise LGTM, weird that they have wheel in install_requires, should be in setup_requires... but that's an upstream issue

**Comment by mmahut** [comment]
Sorry, this felt off my radar. I will try to get it tested with real hardware.

File: pkgs/development/python-modules/ledger_agent/default.nix
Comment (mmahut) [reply pkgs/development/python-modules/ledger_agent/default.nix]: This is a split of the trezor-agent package. I'm fine being listed as maintainer.

File: pkgs/development/python-modules/keepkey_agent/default.nix
Comment (hkjn) [reply pkgs/development/python-modules/keepkey_agent/default.nix]: Let's continue over in the other conversation for `ledger_agent` since it's about the same topic.

File: pkgs/development/python-modules/ledger_agent/default.nix
Comment (hkjn) [reply pkgs/development/python-modules/ledger_agent/default.nix]: Right, I asked @mmahut earlier in [this thread](https://github.com/NixOS/nixpkgs/pull/69153#issuecomment-533889159). I have not gotten an explicit approval from @np that he's fine being listed as a maintainer for all three packages, though figured that since we're splitting an existing package it'd be fine to keep the existing maintainers for the new packages.

**Comment by hkjn** [comment]
Thanks for the review, @jonringer!

Let me know if anyone has any remaining comments that have not been addressed, or otherwise how to progress with getting this PR merged.

**Review by worldofpeace — APPROVED** [review]
LGTM :+1: Build them locally.

File: pkgs/development/python-modules/ledger_agent/default.nix
Comment (jonringer) [reply pkgs/development/python-modules/ledger_agent/default.nix]: sounds good to me then

**Review by jonringer — APPROVED** [review]
`nix-review` passes on NixOS
diff LGTM
executables seem to work

```
[6 built, 271 copied (481.2 MiB), 106.8 MiB DL]
https://github.com/NixOS/nixpkgs/pull/69153
6 package were build:
keepkey_agent ledger_agent python27Packages.ledger_agent python38Packages.keepkey_agent python38Packages.ledger_agent trezor_agent
```

**Comment by jb55** [comment]
awesome, I'll look into rebasing my trezor gpg agent module on this now

**Pull request merged**
